### DISCUSSION TITLE
Offset reconstruction using cone beam FDK

### DISCUSSION THREAD

**Discussion by dominikjohn** [astra-toolbox/astra-toolbox#523] [Q&A]
Hello everyone,

I have a 360° scan where the center of rotation is not in the middle of the detector, but offset to the right. The object extends beyond a single field of view, but all of it is in view at some point during the rotation. Also, it is set up in such a way that the corner of the cylindrical object always stays in the same place during the rotation.

Is it possible to reconstruct this scenario with an FDK? I tried shifting the rotation axis using astra.geom_postalignment, but I always end up with a black volume enclosed by a white circle. I know the offset is roughly 800 pixels.

I'm just asking because the documentation says

> Note: cone_vec geometries that are not relatively close to (non-vec) cone geometries may give unexpected results.

so I was wondering if this already qualifies as "not relatively close to non-vec"?

Thanks a lot in advance!

Best,
Dominik

**Comment by askorikov** [comment]
Hi! If it's just a shift, this should be fine for `geom_postalignment` + FDK. What is meant by "not close to non-vec" is that it wouldn't work for e.g. non-circular scanning trajectories. I have just clarified this in the docs. Instead, the problem may be that the shift is too large (voxel size efffect?), or in a wrong direction. Can you maybe post a code snippet to show what you are doing, and an example projection image if possible?

**Comment by dominikjohn** [comment]
Hi, thank you for the quick reply and for updating the docs! I think I have managed to get a better image that is not just a black circle in the meantime, but it still looks off to me. I have attached the code I'm using and a screenshot (note that I used fewer angles for the preview to make the computation faster). I think the only thing I'm doing differently from the standard FDK case is setting up a volume that is roughly 1.8 times as big as the detector width to accommodate the entire object and using geom_postalignment to try out different offsets.

<img width="630" alt="Bildschirmfoto 2024-11-30 um 09 05 07" src="https://github.com/user-attachments/assets/99069945-7bcd-4c83-87c1-3c7dcdafc40b">

Thanks a lot in advance!

```
vol_geom = astra.create_vol_geom(
    detector_width * 1.8, # added a 2 here to give me more space for the offset reconstruction
    detector_width * 1.8,
    detector_height,
)

scaling_factor = 1e6 # scaling for numerical stability
source_distance = 21.5 * scaling_factor
detector_distance = 0.158 * scaling_factor
pixel_size = 1.444e-6 * scaling_factor

proj_geom = astra.create_proj_geom(
    'cone', 
    pixel_size, 
    pixel_size,
    detector_height,
    detector_width,
    angles,
    source_distance,
    detector_distance
)

proj_geom = astra.geom_2vec(proj_geom)
proj_geom_offset = astra.geom_postalignment(proj_geom, self.center_shift)

# Create ASTRA objects
proj_id = astra.create_projector('cuda3d', proj_geom_offset, vol_geom)
sino_id = astra.data3d.create('-proj3d', proj_geom_offset, chunk_data)
recon_id = astra.data3d.create('-vol', vol_geom)

# Configure reconstruction
cfg = astra.astra_dict('FDK_CUDA')
cfg['ProjectorId'] = proj_id
cfg['ProjectionDataId'] = sino_id
cfg['ReconstructionDataId'] = recon_id
cfg['option'] = {'ShortScan': self.enable_short_scan}

# Run reconstruction
alg_id = astra.algorithm.create(cfg)
astra.algorithm.run(alg_id)
```

**Comment by askorikov** [comment]
These artifacts can be due to numerical overflows I think. Can you check that the input to FDK doesn't have exceptionally large values?
